File: leetcode/tree/102. Binary Tree Level Order Traversal/py/main.py
import collections

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


# Time O(N)
# Space O(N)
class Solution:
    # Time O(N)
    # Space O(N)
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []

        levels = []

        def dfs(node, lvl):
            # start the current level
            if len(levels) == lvl:
                levels.append([])

            # append the current node value
            levels[lvl].append(node.val)

            if node.left:
                dfs(node.left, lvl + 1)

            if node.right:
                dfs(node.right, lvl + 1)

        dfs(root, 0)

        return levels

    # Time O(N)
    # Space O(N)
    def levelOrder_II(self, root: TreeNode) -> List[List[int]]:
        if not root: return []

        nodeQueue = collections.deque()
        lvlQueue = collections.deque()
        prevLvl, lvlList, ans = 0, [], []

        nodeQueue.append(root)
        lvlQueue.append(0)

        while nodeQueue:
            currNode = nodeQueue.popleft()
            currLvl = lvlQueue.popleft()

            if currNode:
                if currLvl == prevLvl + 1:
                    ans.append(lvlList)
                    lvlList = []

                lvlList.append(currNode.val)

                nodeQueue.append(currNode.left)
                nodeQueue.append(currNode.right)

                lvlQueue.append(currLvl + 1)
                lvlQueue.append(currLvl + 1)

                prevLvl = currLvl

        ans.append(lvlList)

        return ans

File: leetcode/tree/102. Binary Tree Level Order Traversal/js/main.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

// Time O(N)
// Space O(N)
const levelOrder = root => {
  if (root === null) {
    return [];
  }

  let ans = [];
  let map = new Map();

  map.set(0, [root.val]);

  function dfs(node, depth) {
    if (!node || !node.children) {
      return;
    }

    let n = node.children.length;

    for (let i = 0; i < n; i++) {
      let item = node.children[i];

      if (map.has(depth)) {
        map.set(depth, [...map.get(depth), item.val]);
      } else {
        map.set(depth, [item.val]);
      }

      if (item.children.length > 0) {
        dfs(item, depth + 1);
      }
    }
  }

  dfs(root, 1);

  for (let [, val] of map) {
    ans.push(val);
  }

  return ans;
};
